Change wxBitmap to wxBitmapBundle in wxMenuItem

kosh543 opened the pull request.

kosh543 committed:
wxBitmapBundle for wxMenuItem on MSW

kosh543 committed:
wxBitmapBundle for wxMenuItem on Mac

kosh543 committed:
wxBitmapBundle for wxMenuItem on GTK

vadz commented on src/gtk/menu.cpp:
```diff
@@ -762,6 +766,36 @@ void wxMenuItem::SetBitmap(const wxBitmap& bitmap)
+        }
+        else
+        {
+            bmp = m_bitmap.GetBitmap(wxDefaultSize);
```
In which case can we get here, i.e. when do we need a bitmap for a menu item without any window?

vadz replied:
I'd still like to know the answer to this.

kosh543 replied:
It can be called from GtkAppend while wxMenu doesn't attached to wxMenubar yet.
Also, it is a public function, so it can be called directly even before wxMenuItem will be attached to wxMenu.

vadz commented on src/gtk/menu.cpp:
```diff
@@ -1110,6 +1139,20 @@ void wxMenu::Attach(wxMenuBarBase *menubar)
+        wxMenuItem *menuitem = node->GetData();
+        if (menuitem->IsSubMenu())
+            menuitem->GetSubMenu()->SetupBitmaps(win);
+        if (menuitem->GetKind() == wxITEM_NORMAL)
```
Do we really need to test for this? AFAICS we could just always call `SetupBitmaps()` and it would do nothing for the others.

It doesn't change anything right now, of course, but if we ever allow bitmaps for check/radio items or add some new menu item kind, it would prevent the bitmaps from being used for it and it could be not obvious to find why.

vadz commented on src/gtk/menu.cpp:
```diff
@@ -762,6 +766,36 @@ void wxMenuItem::SetBitmap(const wxBitmap& bitmap)
     }
 }
 
+wxBitmap wxMenuItem::GetBitmap() const
```
This should ideally reuse the static `wxBitmapBundle::GetBitmapFor()` overload mentioned in another PR discussion.

vadz reviewed (commented):
Thanks, looks good to me.

vadz committed:
Ensure that menu item HBITMAPs live for long enough

We must keep storing the currently used bitmap in wxMSW wxMenuItem to
ensure that its HBITMAP remains valid, as keeping wxBitmapBundle is not
enough: it may not preserve the bitmap at all (if it generates it on
demand, for example), and we may also use a different bitmap from what
we get from it (e.g. we convert mask to alpha).

vadz committed:
Remove wxMenuItem::GetBitmapBundle() from wxGTK

This function doesn't exist in the other ports and is not really needed
in this one neither, so just remove it for now.

If we really want to have it, we need to make it available everywhere.

vadz committed:
Setup bitmaps for all menu item kinds in wxGTK

This doesn't change anything for now, but doesn't cost anything and
might avoid problems in the future if another menu item kind is added.

vadz committed:
Move GetBitmapFromBundle() to wxMenuItemBase

This is just a refactoring to avoid duplicating this function code in
all ports and just have it once in the base class.

vadz committed:
Use scalable icon in the menu sample

Replace the old XPM with a call to wxArtProvider::GetBitmapBundle(),
which looks well even in high DPI.

Also add an item to a popup menu for testing.

vadz commented:
I've fixed the problem with bad `HBITMAP` in wxMSW, addressed some of my own remarks and refactored the code a bit to avoid duplicating it. AFAICS this is ready for merging now, please let me know if you (or anybody else) see any problems with it.

Thanks!
